README.md
# cfsystem

## Description

Configure a bare minimal production system regardless of its purpose. It depends on
more specific [cfnetwork][], [cfauth][] and [cffirehol][] modules.

What it does:

* Whatever [cfnetwork][] does
* Whatever [cfauth][] does
* Whatever [cffirehol][] does
* Setups APT for Debian and Ubuntu
* Setups timezone
* Setups hostname based on certname
* Adds firewall rules as required
* Setups special location/pool facts for hiera lookup (see cfsystem::hierapool below)
* Setups email system
* Setups NTP daemon and command line client for large gap sync
* Setup all locales and the default locale (configurable)
* Manages /etc/profile.d/ & adds allowed bin paths to sudo search_paths
* Installs many handy system tools which almost any admin would expect
* Forces noop scheduler on SSDs and virtual devices (in guests)
* Forces custom I/O scheduler for real spinning HDDs (deadline by default)
* Adds custom rc.local commands, if needed
* Adds cron job to check if running kernel version matches the latest installed (reboot reminder)
* Auto-detect hardware nodes with IPMI
    * Install generic IPMI tools
    * Install Dell-specific tools
    * Other vendors - TODO
* Ruby framework for other cf* modules
* The following helper scripts are installed
    * `cf_clear_email_queue` - clear all emails in exim queue
    * `cf_clear_frozen_emails` - clear only frozen emails in exim queue
    * `cf_send_test_email` - send test email to admin address
    * `cf_kernel_version_check` - check if kernel version mismatch the latest installed one
    * `cf_auto_block_scheduler` - setup auto-detected I/O scheduler per block device
    * `cf_apt_key_updater <key_id>` - run GPG key re-import, if expired
    * `cf_ntpdate` - run pre-configured ntpdate
* Public API for Puppet parser:
    * `Cfsystem::CpuWeight` - cgroup CPU weight
    * `Cfsystem::IoWeight` - cgroup I/O weight
    * `Cfsystem::Keytype` - ssh key types
    * `Cfsystem::Rsabit` - RSA key bits
    * `cfsystem::query` - caching wrapper around `puppetdb_query` (cached per catalog)
    * `cfsystem::stable_sort(arg)` - deep sort of Hash/Array to avoid isomorphic configuration "change"
    * `cfsystem::add_group($user, $group) >> Resource` - make sure user is part of the group
    * `cfsystem::gen_key(name, params, forced_key)` - generate or save persistent SSH key
    * `cfsystem::gen_pass(name, length, forced_pass)` - generate or save persistent password
    * `cfsystem::gen_port(name, forced_port)` - allocate or save persistent network port
    * `cfsystem::pretty_json(data)` - return pretty formatted JSON string
    * `cf_notify` - replacement of standard notify to avoid refresh side-effects


## Technical Support

* [Example configuration](https://github.com/codingfuture/puppet-test)
* Free & Commercial support: [support@codingfuture.net](mailto:support@codingfuture.net)

## Setup

Up to date installation instructions are available in Puppet Forge: https://forge.puppet.com/codingfuture/cfsystem

Please use [librarian-puppet](https://rubygems.org/gems/librarian-puppet/) or
[cfpuppetserver module](https://codingfuture.net/docs/cfpuppetserver) to deal with dependencies.

There is a known r10k issue [RK-3](https://tickets.puppetlabs.com/browse/RK-3) which prevents
automatic dependencies of dependencies installation.

## Examples

Please check [codingufuture/puppet-test](https://github.com/codingfuture/puppet-test) for
example of a complete infrastructure configuration and Vagrant provisioning.

## Implicitly created resources

```yaml
cfnetwork::describe_services:
    puppet:
        server: 'tcp/8140'
    smtp:
        server: 'tcp/25'
    cfsmtp:
        server:
            - 'tcp/25'  # smtp
            - 'tcp/465' # smtps
            - 'tcp/587' # submission
    # if $cfsystem::add_repo_cacher
    'apcng':
        server: 'tcp/3142'
    # if $cfsystem::repo_proxy
    'aptproxy':
        server: "tcp/${proxy_port}"
cfnetwork::service_ports:
    # foreach $cfsystem::email::listen_ifaces
    "${listen_ifaces}:smtp:cfsystem": {}
    'local:smtp:cfsystem': {}
    # if $cfsystem::add_ntp_server
    "${cfsystem::service_face}:ntp": {}
    # if $cfsystem::add_repo_cacher
    "${cfsystem::service_face}:apcng:cfsystem": {}
    # if ${cfsystem::service_face} not in ['any', 'local']
    'local:apcng:cfsystem': {}
cfnetwork::client_ports:
    'any:puppet:cfsystem':
        user: 'root'
    'local:smtp:cfsystem': {}
    # if $smarthost = undef then dst filtering is disabled
    'any:cfsmtp:cfsystem':
        user => ['root', 'Debian-exim'],
        dst  => $smarthost
    'any:ntp:cfsystem':
        user => ['root', 'ntpd'],
    # if $cfsystem::add_repo_cacher
    'any:http:apcng':
        user: 'apt-cacher-ng'
    # if $cfsystem::add_repo_cacher
    'any:https:apcng':
        user: 'apt-cacher-ng'
    # if $cfsystem::repo_proxy
    'any:aptproxy:cfsystem':
        dst: $proxy_host
        user: 'root'
    # if not $cfsystem::repo_proxy
    'any:http:cfsystem':
        user: 'root'
    # if not $cfsystem::repo_proxy
    'any:https:cfsystem':
        user: 'root'
```

## Class parameters

## `cfsystem` class

* `allow_nfs = false` - purge RPC packages unless true
* `admin_email = undef` - email address to use for `root` and as the default sink
* `repo_proxy = undef` - if set, use the config as HTTP/HTTPS proxy for package retrieval.
    * `host` - IP or hostname
    * `port` - TCP port
* `add_repo_cacher = false` - if true, install apt-cacher-ng and accept clients on `$service_face`
* `service_face = 'any'` - interface to accept client for NTP and HTTP proxy, if enabled separately
* `ntp_servers = [ '0.pool.ntp.org', '1.pool.ntp.org', '2.pool.ntp.org', '3.pool.ntp.org' ]` - upstream NTP server
* `add_ntp_server = false` - if true, accept NTP service clients on `$service_face`
* `Enum['ntp', 'openntpd', 'chrony', 'systemd'] $ntpd_type = 'systemd'` - NTP implementation to use
* `timezone = 'Etc/UTC'` - setup system timezone
* `apt_purge` - passed to apt::purge, purge all sources and preferences by default
* `apt_update` - passed to apt::update, update daily with 300 second timeout by default
* `apt_pin = 1001` - default priority (>=1001 - force downgrades to make the system consistent)
* `apt_backports_pin = 600` - default priority (>=1001 - force downgrades to make the system consistent)
* `real_hdd_scheduler` - default scheduler for not SSD and not virtualized HDDs
* `rc_local` - list of additional commands to add to /etc/rc.local
    (SSD and virtual is always 'noop')
* `puppet_host = "puppet.${::trusted['domain']}"` - Puppet Server hostname
* `puppet_cahost = $puppet_host` - Puppet CA hostname
* `puppet_env = $::environment` - Puppet environment
* `puppet_use_dns_srv = false` - enable support DNS SRV records instead of hostnames
* `mcollective = false` - controls if mcollective service is enabled
* `locale = 'en_US.UTF-8'` - default system locale
* `reserve_ram` = 64 - amount of ram to reserve for system in automatic calculations
* `$key_server = 'hkp://pgp.mit.edu:80'` - default PGP key server
* `$random_feed = true` - enable random entropy generating daemon
* `$add_handy_tools = true` - install additional tools
* `$puppet_backup_age = '1d'` - how long to keep local puppet filebucket backups

## `cfsystem::bindir` type

Setup /etc/profile.d/ & /etc/sudoers.d/ entries for trusted global bin paths. It should not be
configured by user. It's API for other modules.

* `bin_dir` - absolute path to directory for global search path

## `cfsystem::hierapool` class

Automatically including by `cfsystem`. This values are useful in hiera.yaml configuration
to setup hierarchy based on location and tenant/server pool in it. Example:

```yaml
    ---
    :backends:
    - yaml
    :hierarchy:
    - "%{::trusted.domain}/%{::trusted.hostname}"
    - "%{::trusted.domain}"
    - "%{::cf_location}/%{::cf_location_pool}"
    - "%{::cf_location}"
    - common
    :merge_behavior: deeper
    :yaml:
    :datadir:
```

* `location = undef` - if set, saved into `/etc/cflocation`
* `pool = undef` - if set, aved into `/etc/cflocationpool`


## `cfsystem::email` class

Setup email server for outgoing emails. **Please not that this configuration
is not intended to accept internet traffic.**

* `smarthost = undef` - if set, use as smarthost to relay outgoing emails through
* `smarthost_login = undef` - if set, use as login on smarthost
* `smarthost_password = undef` -  if set, use as password on smarthost (plain text)
* `relay_nets = <private subnets>` - allowed clients for SMTP relay, if relay is enabled
    with `$listen_ifaces`
* `listen_ifaces = undef` - list of interface (`cfnetwork::iface` names), besides `lo` to
    listen for SMTP client relay
* `disable_ipv6 = true` - if true, IPv6 supports gets disabled (most likely you
    need it disabled for SMTP)

## `cfsystem::sysctl` class

Setup sysctl entries.

* `vm_swappiness = 1` - 0-100 (%) minimize swap activity by default
* `vm_mmax_map_count = 262144` - increased by default

## `cfsystem::debian` class

Debian-specific configuration.

* `apt_url = 'http://deb.debian.org/debian'` - APT base URL for Debian repos
* `security_apt_url = 'http://security.debian.org/'` - APT base URL for Debian security repo
* `release` = 'jessie' - Debian release name to configure

## `cfsystem::ubuntu` class

Ubuntu-specific configuration.

* `apt_url = 'mirror://mirrors.ubuntu.com/mirrors.txt'` - APT base URL for Ubuntu repos
* `release = 'wily'` - Ubuntu release name to configure

## `cfsystem::debian::debconf` type

* `package = $title` - package to configure & install
* `ensure = present` - passed to `package ensure`
* `config = []` - config entries for `debconf-set-selections`

## `cfsystem::apt::key` type

Configure APT key & add automatic update of expired keys

* `id` - PGP key ID
* `extra_opts = {}` - any additional options for `apt::key`

## `cfsystem::dotenv` type

A special helper to create entries in user ~/.env files

* `user` - previously defined user{ $user: home => 'path'} ($home must be explicitly set)
* `variable` - variable name
* `value` - value
* `env_file = '.env'` - name of .env file relative to $home

## `cfsystem::puppetpki` type
Make actual Puppet PKI (CA, CRL, client cert and private key) data available to specific user.
By default the data is copied under ~/pki/puppet/.

* `user = $title` - local user to use
* `copy_key = true` - if true then private keys of local machine are copied as well
* `pki_dir = undef` - override the default destination folder

## `cfsystem::haproxy` class

Setup haproxy package. No configuration. Used by other modules

* `$disable_standard = true` - controls if default HAProxy service must be disabled

## `cfsystem::randomfeed` class

Setup random entropy generating tools

* `$type = 'haveged'` - tools type
* `$threshold = 2048` - minimal random entropy level

## `cfsystem::sshdir` type

Create a basic `~/.ssh/` directory for unattended user account.

* `$user = $title` - system user with 'home' parameter

## `cfsystem::clusterssh` type

This feature is trade-off between SSH setup in cluster and security. This functionality
creates a single SSH client key and shares across all nodes in cluster. It helps to
get rid of puppet facts processing for target-generated secrets.

Besides shared private key, another problem is clear-text private key getting into
puppet catalog (which should be secured as well).

* `$namespace` - cluster namespace, e.g. 'cfdb'
* `$cluster` - cluster identifier
* `$is_primary` - controls if a new key can be generated
* `$peer_ipset` - name of pre-defined cfnetwork::ipset
* `$user` - system user with 'home' parameter
* `$group = $user` - the user's group
* `$key_type = 'ed25519'` - SSH key type
* `$key_bits = 2048` - SSH key bits (for RSA)

## `cfsystem::hwm` class

Generic class for HardWare Management

* `Enum['none', 'auto', 'generic', 'dell', 'smc'] $type = 'auto'` - 
    select type of HW vendor, if auto-detection fails.

## `cfsystem::hwm::generic` class

Just a placeholder for generic IPMI system.

## `cfsystem::hwm::dell` class

Support for Dell PowerEdge family.

* `$community_repo = 'http://linux.dell.com/repo/community'`

## `cfsystem::hwm::smc` class

Placeholder for SuperMicro support. Not implemented yet.

## `cfsystem::pip` class

Setup latest pip for Python 2&3 into /usr/local.

## `cfsystem::metric` type

Mostly for internal purposes to declare items for cfmetrics monitoring.

## `cf_notify` type

The standard `notify` type has a side effect - it generates refresh event
what may harm automation which expects 0 exit code on no resource changes.
Therefore, this drop-in replacement has been provided.

* `message = $title` - message to show

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. This
project adheres to [Semantic Versioning](http://semver.org/).

## 1.3.2 (2019-11-13)
- FIXED: cf_apt_key_updater for Ubuntu 18.04
- FIXED: to enable timers on startup
- NEW: additional system startup commands

## 1.3.1 (2019-06-17)
- FIXED: to allow cftimer- service tuning through settings_tune
- FIXED: to allow cfsvc- service tuning through settings_tune
- NEW: allow borrowing insufficient memory from swap for distribution
- NEW: implemented local backup filebucket cleanup

## 1.3.0 (2019-04-14)
- FIXED: forced secrets to override persistent values
- FIXED: atomicWrite to obey ownership & mode when content is the same
- FIXED: cfsystem_timer exceptions in some configurations
- CHANGED: to prune /opt/codingfuture/bin
- CHANGED: Ubuntu instance to remove update-motd package
- CHANGED: got rid of historical cf-apt-update workaround
- NEW: 'silent' atomic file write API for temporary cases

## 1.2.0 (2018-12-09)
- CHANGED: updated for Ubuntu 18.04 Bionic support
- CHANGED: to use cfhttp service in firewall config
- CHANGED: enabled cgroup-v2 for kernels 4.5+
- NEW: cfsystem::add_handy_tools to control additional package setup
- NEW: FreeIPA NTP support
- NEW: cfsystem_service type
- NEW: cfsystem_timer type

## 1.1.0 (2018-05-02)
- NEW: cfsystem::metric type as a sort of IoC to integrate cfmetrics
- NEW: Copy-on-Write reserve feature for service definition (overcommit)
- NEW: cfsystem::add_group functions

## 1.0.2 (2018-04-29)
- CHANGED: to allow zero min/max memory requirements
- NEW: cfsystem::pip class for latest pip setup in /usr/local

## 1.0.1 (2018-04-13)
- FIXED: Metaspace JVM parameter detection with JRE 1.8.0_162+
- FIXED: updated exim4 config template

## 0.12.9 (2018-03-24)
- NEW: generic cfsystem::sshdir with custom configuration extension support

## 0.12.8 (2018-03-19)
- NEW: cf_notify as replacement for standard notify to avoid its refresh side effects
- NEW: added tshark package for standard installation

## 0.12.7 (2018-03-15)
- CHANGED: to always prefer PSON serialization to avoid retries
- CHANGED: moved syslog related stuff solely to cflogsink

## [0.12.6](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.12.6)
- CHANGED: improved APT cache, Email & NTP service security with ipset:localnet
- FIXED: internal API to allow custom EnvironmentFile for systemd unit
- NEW: cfsystem::location variable
- NEW: cfsystem::netsyslog functionality
- NEW: cfsystem::hdsyslog functionality
- NEW: CfSystem.makeVersion now supports directory traversal
- NEW: show amount of unused RAM in memory distribution

## [0.12.5](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.12.5)
- CHANGED: to mask instead of just disable agent/mcollective, if required

## [0.12.4](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.12.4)
- FIXED: support systemd 236+ timesyncd setup

## [0.12.3](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.12.3)
- CHANGED: to use systemd-timesyncd by default
- NEW: systemd-timesyncd support
- NEW: cfsystem::sysctl::vm_mmax_map_count
- NEW: custom systemd services to use service name for syslog tagging

## [0.12.2](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.12.2)
- FIXED: minor ntp.conf configuration issues

## [0.12.1](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.12.1)
- CHANGED: cfsystem::service_face to default to cfnetwork::service_face
- FIXED/CHANGED: ntpd configuration to use "server" instead of "pool"

## [0.12.0](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.12.0)
- CHANGED: to use Puppet 5 by default
- FIXED: Puppet 5 runtime issues
- FIXED: cf_auto_block_scheduler to work with stricter "test"

## [0.11.9](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.11.9)
- NEW: cfsystem::clusterssh adds also public key file for convenience

## [0.11.8](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.11.8)
- FIXED: clusterssh / PuppetX::CfSystem::Util.genKeyCommon to force new keys
    on secondary instances when old one is already set

## [0.11.7](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.11.7)
- FIXED: kernel version check to filter out only installed versions

## [0.11.6](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.11.6)
- FIXED: kernel version check to use natural sorting after version extraction

## [0.11.5](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.11.5)
- FIXED: kernel version check to use natural sorting

## [0.11.4](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.11.4)
- CHANGED: to use now available Stretch repository

## [0.11.3](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.11.3)
- CHANGED: LimitMEMLOCK is set to "infinity" instead of MemoryMax size, if requested
- NEW: absolute cfsystem::dotenv filename support
- NEW: string support to systemd memory limits (internal API)

## [0.11.2](https://github.com/codingfuture/puppet-cfsystem/releases/tag/v0.11.2)
- FIXED: to allow execute cf_wait_socket helper by any user
- CHANGED: to use http://deb.debian.org/debian as default for Debian
- CHANGED: to use $apt_backports_pin = 1001 by default for old system
- CHANGED: updated to APT module >= 4.1
- NEW: Puppet 5.x support
- NEW: Ubuntu Zesty support
- NEW: cfsystem::pretty_json

## [0.11.1]
- Changed ntpdate cron to mail output only if |time offset| >= 1

## [0.11.0]
- Minor fix for HAProxy setup @ jessie
- Fixed to also pin libssl-dev @ jessie
- Fixed to properly create slice extension
- Updated to new 'cfnetwork::bind_address' API
- Added cf_ntpdate wrapper & cron job
- Changed not to run apt-update only when require, but not daily
    - Solves cases of failed APT update during catalog deploy
    - Added second attempt on failure
- Fixed a long standing issues with "exists" in ensure processing

lib/puppet_x/cf_system.rb
#
# Copyright 2016-2019 (c) Andrey Galkin
#


require 'puppet/util/logging'
require 'puppet_x'
require 'puppet/util/diff'
require 'puppet/util/execution'
require 'fileutils'
require 'securerandom'
require 'find'

# Done this way due to some weird behavior in tests also ignoring $LOAD_PATH
require File.expand_path( '../cf_system/config', __FILE__ )

module PuppetX::CfSystem
    CFSYSTEM_CONFIG = '/etc/cfsystem.json'
    CUSTOM_BIN_DIR = '/opt/codingfuture/bin'
    SYSTEMD_DIR = '/etc/systemd/system'
    SYSTEMD_CTL = '/bin/systemctl'
    SUDO = '/usr/bin/sudo'
    WAIT_SOCKET_BIN = "#{CUSTOM_BIN_DIR}/cf_wait_socket"
    
    
    #---
    BASE_DIR = File.expand_path('../', __FILE__)
    require "#{BASE_DIR}/cf_system/provider_base"
    require "#{BASE_DIR}/cf_system/util"
    
    class << self
        attr_accessor :config
        attr_accessor :memory_distribution
        
        include Puppet::Util::Logging
        Puppet::Util.logmethods(self, true)
    end
    
    def self.fileVersionTraverse(digest, file)
        if file.is_a? Array
            file.each do |f|
                fileVersionTraverse(digest, f)
            end
        elsif FileTest.directory?(file)
            Find.find(file) do |p|
                if FileTest.directory?(p)
                    next
                else
                    digest.file(p)
                end
            end
        else
            digest.file(file)
        end
    end

    def self.makeVersion(file)
        digest =Digest::MD5.new
        fileVersionTraverse(digest, file)    
        digest.hexdigest
    end
    
    def self.begin
        debug('cfsystem::begin')
        self.config = Config.new(CFSYSTEM_CONFIG)
        File.exists?(CFSYSTEM_CONFIG)
    end
    
    def self.commit
        debug('cfsystem::commit')
        self.config.save()
        self.config = nil
        self.post_commit()
    end
    
    def self.post_commit
        res = Puppet::Util::Execution.execute(
            ['/opt/codingfuture/bin/cf_kernel_version_check'],
            {
                :failonfail => false,
                :combine => true,
            }
        )
        warning(res) if res.exitstatus != 0
    end
    
    def self.atomicWrite(file, content, opts={})
        content = content.join("\n") + "\n" if content.is_a? Array

        mode = opts.fetch(:mode, 0600)
        user = opts.fetch(:user, 'root')
        group = opts.fetch(:group, user)

        if File.exists?(file) and (content == File.read(file))
            debug("Content matches for #{file}")
            FileUtils.chmod(mode, file)
            FileUtils.chown(user, group, file)
            return false
        end
        
        return true if opts[:dry_run]
        
        #---
        tmpfile = file + ".#{$$}"
        
        File.open(tmpfile, 'w+', mode) do |f|
            f.write(content)
        end
        
        # Show diff
        #---
        if opts.fetch(:silent, false)
            # pass
        elsif opts.fetch(:show_diff, true)
            if File.exists?(file)
                diff = Puppet::Util::Diff.diff(file, tmpfile)
                opts.fetch(:mask_diff, []).each do |v|
                    diff.gsub!(v, '<secret>')
                end
                notice("File[#{file}]/content:\n" + diff)
            else
                notice("File[#{file}]/content:\n" + content)
            end
        else
            notice("File[#{file}] hidden content change")
        end

        # Atomically move config file to its location
        #---
        FileUtils.chown(user, group, tmpfile)
        File.rename(tmpfile, file)
        debug("Writed a new #{file}")
        
        return true
    end
    
    def self.atomicWriteIni(file, settings, opts={})
        content = []
        settings.each do |section, subsettings|
            content << "[#{section}]"
            subsettings.each do |k, v|
                if v.is_a? Array
                    v.each do |ve|
                        content << "#{k}=#{ve}"
                    end
                else
                    content << "#{k}=#{v}"
                end
            end
            content << ''
        end
        
        content = content.join("\n")
        
        self.atomicWrite(file, content, opts)
    end
    
    def self.atomicWriteEnv(file, settings, opts={})
        content = []
        settings.each do |k, v|
            if v.is_a? Array
                v = v.join(' ')
            end
            
            content << "#{k}=\"#{v}\""
        end
        
        content << ''
        
        content = content.join("\n")
        
        self.atomicWrite(file, content, opts)
    end
    
    def self.calcMemorySections()
        debug('Calculating RAM sections')

        memory_fact = Facter['memory']
        total_ram = memory_fact.value['system']['total_bytes']
        total_ram = total_ram / 1024 / 1024
        total_swap = memory_fact.value.fetch('swap', {'total_bytes' => 0})['total_bytes']
        total_swap = total_swap / 1024 / 1024
        debug("Total RAM: #{total_ram}MB SWAP:#{total_swap}MB")
        
        newconf = self.config.get_new('memory_weight')
        
        # Calc minimum allocations and totals
        #---
        min_ram = 0
        total_weight = 0
        min_weight = 0
        merged_conf = {}
        
        newconf.each do |k, v|
            weight = v[:weight]
            min_mb = v[:min_mb]
            total_weight += weight
            
            if not min_mb.nil?
                min_ram += min_mb
                min_weight += weight
            end
            
            # Support sub-section definition
            k = k.split('/')[0]
            
            if merged_conf.has_key? k
                mv = merged_conf[k]
                mv[:weight] += weight
                
                min_mb = ((mv[:min_mb] || 0) + (min_mb || 0))
                max_mb = ((mv[:max_mb] || 0) + (v[:max_mb] || 0))
                mv[:min_mb] = min_mb if min_mb > 0
                mv[:max_mb] = max_mb if max_mb > 0
            else
                merged_conf[k] = v.clone
            end
        end
        
        if total_ram < min_ram
            if (total_ram + total_swap) >= min_ram
                # Allow to borrow swap, but do not distribute that normally
                warning("Borrowing #{min_ram - total_ram}MB from swap!")
                total_ram = min_ram
            else
                warning('Memory sections: ' + newconf.to_s )
                raise ArgumentError, "Minimal ram required #{min_ram} is more than available #{total_ram}"
            end
        end
        
        # Pass #1 whole memory distribution
        #---
        avail_ram = total_ram - min_ram
        total_min_ratio = (1 - min_ram.to_f / total_ram)
        alloc_ram = 0
        mem_distrib = {}
        
        merged_conf.each do |k, v|
            min_mb = v[:min_mb]
            min_mb = 0 if min_mb.nil?
            max_mb = v[:max_mb]
            weight = v[:weight] 
            
            calc_mem = (total_ram * weight * total_min_ratio / total_weight).to_i
            calc_mem += min_mb

            # Well, it's not completely fair situation
            if not max_mb.nil? and calc_mem > max_mb
                calc_mem = max_mb
            end
                
            alloc_ram += calc_mem
            mem_distrib[k] = calc_mem
        end
        
        # Pass #2 unallocated memory distribution
        #---
        unalloc_ram = total_ram - alloc_ram
        merged_conf.each do |k, v|
            next if not v[:max_mb].nil?
            
            weight = v[:weight] 
            
            calc_mem = (unalloc_ram * weight / total_weight).to_i
            alloc_ram += calc_mem
            mem_distrib[k] += calc_mem
        end

        mem_distrib['free'] = unalloc_ram if alloc_ram < total_ram
        
        #---
        self.memory_distribution = mem_distrib
        notice('Memory distribution result: ' + mem_distrib.to_s )
    end
    
    def self.getMemory(name)
        self.memory_distribution[name]
    end
    
    def self.genPort(assoc_id, forced_port=nil)
        ports = self.config.get_persistent('ports')
        PuppetX::CfSystem::Util.genPortCommon(ports, assoc_id, forced_port)
    end
        
    def self.genSecret(assoc_id, len=24, set=nil)
        secrets = self.config.get_persistent('secrets')
        PuppetX::CfSystem::Util.genSecretCommon(secrets, assoc_id, len, set)
    end
        
   
    def self.fitRange(min, max, val=nil)
        val = max if val.nil?
        return [min, [max, val].min].max
    end
    
    def self.roundTo(to, val)
        return (((val + to) / to).to_i * to).to_i
    end

    def self.createTimer(options)
        timer_name = options[:service_name]
        timer_file = "#{SYSTEMD_DIR}/#{timer_name}.timer"
        period = options[:period]
        calendar = options[:calendar]

        content_ini = options.fetch(:content_ini, {})

        # Unit
        #---
        content_ini['Unit'] = {} unless content_ini.has_key? 'Unit'
        unit_ini = content_ini['Unit']
        unit_ini['Description'] ||= timer_name
        #unit_ini['DefaultDependencies'] ||= 'no'

        # Install
        #---
        content_ini['Install'] = {
            'WantedBy' => 'multi-user.target',
        } unless content_ini.has_key? 'Install'

        # Timer
        #---
        content_ini['Timer'] = {} unless content_ini.has_key? 'Timer'
        timer_ini = content_ini['Timer']

        unless period.nil?
            timer_ini['OnBootSec'] = period
            timer_ini['OnUnitActiveSec'] = period
        end

        unless calendar.nil?
            timer_ini['OnCalendar'] = calendar
            timer_ini['Persistent'] = 'yes'
        end

        reload = atomicWriteIni(timer_file, content_ini,
                                {:mode => 0644,
                                 :dry_run => options[:dry_run]})

        # reload on demand
        #---
        if reload
            Puppet::Util::Execution.execute([SYSTEMD_CTL, 'daemon-reload'])
        end

        Puppet::Util::Execution.execute([SYSTEMD_CTL, 'enable', "#{timer_name}.timer"])

        return reload
    end

    def self.createLimitsCommon(section_ini, options)
        cpu_weight = options.fetch(:cpu_weight, nil)
        io_weight = options.fetch(:io_weight, nil)
        mem_limit = options.fetch(:mem_limit, nil)
        cow_reserve = options.fetch(:cow_reserve, 0)
        mem_lock = options.fetch(:mem_lock, false)
        cgroups_v2 = Gem::Version.new(Facter['kernelmajversion'].value()) >= Gem::Version.new('4.5')
        
        mem_limit = "#{mem_limit + cow_reserve}M" if mem_limit.is_a? Integer
        
        if cgroups_v2
            unless cpu_weight.nil?
                section_ini['CPUAccounting'] = 'true'
                cpu_shares = (10000 * cpu_weight.to_i / 100).to_i
                section_ini['CPUWeight'] = fitRange(1, 10000, cpu_shares)
            end
            
            unless io_weight.nil?
                io_weight = (10000 * io_weight.to_i / 100).to_i
                section_ini['IOAccounting'] = 'true'
                section_ini['IOWeight'] = fitRange(1, 10000, io_weight)
            end
            
            unless mem_limit.nil?
                section_ini['MemoryAccounting'] = 'true'
                section_ini['MemoryHigh'] = mem_limit
                section_ini['MemoryMax'] = "#{(mem_limit.to_i * 1.5).to_i}M"
                        
                if mem_lock
                    section_ini['LimitMEMLOCK'] = "infinity"
                end
            end
        else
            unless cpu_weight.nil?
                section_ini['CPUAccounting'] = 'true'
                cpu_shares = (1024 * cpu_weight.to_i / 100).to_i
                section_ini['CPUShares'] = fitRange(2, 262144, cpu_shares)
            end
            
            unless io_weight.nil?
                io_weight = (500 * io_weight.to_i / 100).to_i
                section_ini['BlockIOAccounting'] = 'true'
                section_ini['BlockIOWeight'] = fitRange(10, 1000, io_weight)
            end
            
            unless mem_limit.nil?
                section_ini['MemoryAccounting'] = 'true'
                section_ini['MemoryLimit'] = mem_limit
                        
                if mem_lock
                    section_ini['LimitMEMLOCK'] = "infinity"
                end
            end
        end
    end
    
    def self.createSlice(options)
        slice_name = options[:slice_name]
        slice_file = "#{SYSTEMD_DIR}/#{slice_name}.slice"
        
        content_ini = options.fetch(:content_ini, {})
        
        # Unit
        #---
        content_ini['Unit'] = {} unless content_ini.has_key? 'Unit'
        unit_ini = content_ini['Unit']
        unit_ini['Description'] ||= slice_name
        unit_ini['DefaultDependencies'] ||= 'no'
        unit_ini['Before'] ||= 'slices.target'

        # Service
        #---
        content_ini['Slice'] = {} unless content_ini.has_key? 'Slice'
        slice_ini = content_ini['Slice']
        
        self.createLimitsCommon(slice_ini, options)
        
        reload = atomicWriteIni(slice_file, content_ini,
                                {:mode => 0644,
                                 :dry_run => options[:dry_run]})
        
        # reload on demand
        #---
        if reload
            Puppet::Util::Execution.execute([SYSTEMD_CTL, 'daemon-reload'])
        end
        
        return reload
    end
    
    def self.createService(options)
        service_name = options[:service_name]
        user = options.fetch(:user, service_name)
        group = options.fetch(:group, user)
        
        service_file = "#{SYSTEMD_DIR}/#{service_name}.service"
        env_file = "/etc/default/#{service_name}.conf"
        
        content_ini = options[:content_ini].clone
        content_env = options.fetch(:content_env, {}).clone
        
        # Unit
        #---
        content_ini['Unit'] = {} unless content_ini.has_key? 'Unit'
        unit_ini = content_ini['Unit']
        unit_ini['Description'] ||= service_name
        unit_ini['After'] ||= 'syslog.target network.target'
        
        # Install
        #---
        content_ini['Install'] = {
            'WantedBy' => 'multi-user.target',
        } unless content_ini.has_key? 'Install'

        # Service
        #---
        service_ini = content_ini['Service']
        service_ini.replace({
            'SyslogIdentifier' => service_name,
            'Type' => 'simple',
            'Restart' => 'always',
            'RestartSec' => 5,
            'User' => user,
            'Group' => group,
            'UMask' => '0027',
            'RuntimeDirectory' => service_name,
        }.merge service_ini)

        unless content_env.empty?
            if !service_ini['EnvironmentFile']
                service_ini['EnvironmentFile'] = env_file
            elsif service_ini['EnvironmentFile'].is_a? Array
                service_ini['EnvironmentFile'] << env_file
            else
                service_ini['EnvironmentFile'] = [
                    service_ini['EnvironmentFile'],
                    env_file,
                ]
            end
        end
        
        self.createLimitsCommon(service_ini, options)
        
        # Write configs
        #---
        if content_env.empty?
            FileUtils.rm_f(env_file)
        else
            env_changed = atomicWriteEnv(env_file, content_env, {:mode => 0644})
        end
        
        reload = atomicWriteIni(service_file, content_ini, {:mode => 0644})
        
        # reload on demand
        #---
        if reload
            Puppet::Util::Execution.execute([SYSTEMD_CTL, 'daemon-reload'])
        end
        
        # Make sure it's enabled
        Puppet::Util::Execution.execute([SYSTEMD_CTL, 'enable', service_name, '--no-reload'])
        
        return env_changed || reload
    end
    
    def self.maskService(service_name)
        service_file = "#{SYSTEMD_DIR}/#{service_name}.service"
        
        if !File.exists?(service_file) or !File.symlink?(service_file)
            warning("> systemd masking #{service_name}")
            Puppet::Util::Execution.execute([SYSTEMD_CTL, 'stop', service_name])
            Puppet::Util::Execution.execute([SYSTEMD_CTL, 'mask', service_name])
        end
    end
    
    def self.cleanupSystemD(prefix, new_files, ext='service')
        old_files = Dir.glob("#{SYSTEMD_DIR}/#{prefix}*.#{ext}").
                            map { |v| File.basename(v, ".#{ext}") }
        old_files -= new_files
        old_files.each do |s|
            file = "#{s}.#{ext}"
            
            if ext == 'service'
                begin
                    warning("Stopping old systemd file: #{file}")
                    Puppet::Util::Execution.execute([SYSTEMD_CTL, 'stop', file])
                rescue => e
                    err("Fail on stop: #{e}")
                end
            end
            
            warning("Removing old systemd file: #{file}")
            FileUtils.rm_f "#{SYSTEMD_DIR}/#{file}"
        end
        
        if old_files.size
            Puppet::Util::Execution.execute([SYSTEMD_CTL, 'daemon-reload'])
        end
    end
end
